src/clients/UBAClient/UBAClientAbstract.ts

```diff
@@ -0,0 +1,224 @@
+   * @note be exposed to finality risk. Additional verification that can only be performed within the UBA context:
+   * @note - Only the first instance of a partial fill for a deposit is accepted. The total deposit amount is taken, and
+   * @note   subsequent partial, complete or slow fills are disregarded.
+   * @param spokePoolClient SpokePoolClient instance for this chain.
+   * @param fromBlock       Optional lower bound of the search range. Defaults to the SpokePool deployment block.
+   * @param toBlock         Optional upper bound of the search range. Defaults to the latest queried block.
+   */
+  public abstract getFlows(chainId: number, fromBlock?: number, toBlock?: number): UbaFlow[];
+
+  /**
+   * @description Evaluate an RefundRequest object for validity.
+   * @dev  Callers should evaluate 'valid' before 'reason' in the return object.
+   * @dev  The following RefundRequest attributes are not evaluated for validity and should be checked separately:
+   * @dev  - previousIdenticalRequests
+   * @dev  - Age of blockNumber (i.e. according to SpokePool finality)
+   * @param chainId       ChainId of SpokePool where refundRequest originated.
+   * @param refundRequest RefundRequest object to be evaluated for validity.
+   */
+  public abstract refundRequestIsValid(
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    _chainId: number,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    _refundRequest: RefundRequestWithBlock
+  ): RequestValidReturnType;
+
+  protected abstract instantiateUBAFeeCalculator(chainId: number, token: string, fromBlock: number): Promise<void>;
+
+  /**
+   * Calculate the balancing fee of a given token on a given chainId at a given block number
+   * @param spokePoolToken The token to get the balancing fee for
+   * @param amount The amount to get the balancing fee for
+   * @param hubPoolBlockNumber The block number to get the balancing fee for
+   * @param depositChainId The chainId of the deposit
+   * @param refundChainId The chainId of the refund
+   * @returns The balancing fee for the given token on the given chainId at the given block number
+   */
+  protected async computeBalancingFee(
+    spokePoolToken: string,
+    amount: BigNumber,
+    hubPoolBlockNumber: number,
+    depositChainId: number,
+    refundChainId: number
+  ): Promise<BalancingFeeReturnType> {
+    // Verify that the spoke clients are instantiated.
+    await Promise.all([
+      this.instantiateUBAFeeCalculator(depositChainId, spokePoolToken, hubPoolBlockNumber),
+      this.instantiateUBAFeeCalculator(refundChainId, spokePoolToken, hubPoolBlockNumber),
+    ]);
+    // Get the balancing fees.
+    const [{ balancingFee: depositBalancingFee }, { balancingFee: refundBalancingFee }] = await Promise.all([
+      this.spokeUBAFeeCalculators[depositChainId][spokePoolToken].getDepositFee(amount),
```

Comment:
Do you always need to compute both the deposit and refund balancing fees for any given `amount`? I thought you'd only need to compute one, based on which direction the flow is

Reply:
Good call - updated: https://github.com/across-protocol/sdk-v2/pull/236/commits/65cb9a0ebe0e99e47ff2d71bee03fa486ab54fc1